# Calculate error rates from Sep/17-Sep/21 in office testing. 
Testing build Survey2026 build #70 7192ca2. Built on 2026-09-16T16:13:20Z
https://github.com/ishizuki-tech/Survey2026/releases?page=1#release-build-70-7192ca2

In [13]:
import json

In [14]:
#define the path to the reference text
# One entry per survey recording: survey_id, question_id, expected_answer, wav_file, local_wav
reference_file = "office_testing/expected_answers.json"
reference = reference_file

# Read from a relative path
with open(reference, 'r', encoding='utf-8') as g:
    ref_data = json.load(g)

# Look up the expected answer by (survey_id, question_id), since a survey can have its own reference text
ref_lookup = {(r["survey_id"], r["question_id"]): r["expected_answer"] for r in ref_data}

print (f"Reference data: loaded {len(ref_data)} entries from {reference_file}")
for r in ref_data:
    print(f"{r['survey_id'][:8]} {r['question_id']}: {r['expected_answer']}")

Reference data: loaded 30 entries from office_testing/expected_answers.json
46ec4547 Q8: Ndiyo, iliharibu majani na kupunguza mavuno.
46ec4547 Q9: Naweza kukubali kupoteza hadi asilimia kumi ya mavuno.
46ec4547 Q10: Ikiwa uharibifu unafika karibu asilimia thelathini, nitabadilisha aina.
46ec4547 Q11: Iwe na mavuno mengi na iwe inastahimili ukame na wadudu.
46ec4547 Q12: Hata tani moja kwa ekari, kama mbegu ni nzuri.
46ec4547 Q13: Wakati wa maua na kutengeneza punje.
46ec4547 Q14: Ninaweka ya chakula, nauza iliyobaki na kidogo ni ya mifugo.
46ec4547 Q15: Nawapa kuku na nguruwe mahindi mara mbili hadi tatu kwa wiki. Nawapa kuku na nguruwe mahindi mara mbili hadi tatu kwa wiki.
46ec4547 Q16: Ninanunua kwa agrovet kwa sababu ninaamini ubora wake.
46ec4547 Q17: Nauza kwa wafanyabiashara wa eneo kwa sababu wako karibu na hulipa haraka.
4f629181 Q8: Ndiyo, iliharibu majani na kupunguza mavuno.
4f629181 Q9: Naweza kukubali kupoteza hadi asilimia kumi ya mavuno.
4f629181 Q10: Ikiwa uharibifu un

In [15]:
# Define the GitHub URLs of the surveys we want to test today.
# Paste the normal github.com "blob" links; they are converted to raw file URLs below
survey_urls = [
    "https://github.com/ishizuki-tech/SurveyExports/blob/main/2026-09-17/exports/survey_46ec4547-5bbe-4b02-9c2a-cbcec7223e2c_2026-09-17_16-00-23.json",
    "https://github.com/ishizuki-tech/SurveyExports/blob/main/2026-09-21/exports/survey_4f629181-d1dc-46cf-8e71-d819a4b33662_2026-09-21_14-37-06.json",
    "https://github.com/ishizuki-tech/SurveyExports/blob/main/2026-09-21/exports/survey_7db59ab6-e13d-4080-bb07-3842d8b51fe4_2026-09-21_08-45-52.json",
]

from urllib.request import urlopen

# One entry per survey: file name plus the reference/answer pairs that matched.
surveys = []
for survey_url in survey_urls:
    raw_url = survey_url.replace("https://github.com/", "https://raw.githubusercontent.com/", 1).replace("/blob/", "/", 1)
    with urlopen(raw_url) as f:
        survey_data = json.load(f)

    survey_file = raw_url.rsplit("/", 1)[-1]
    print(f"\nLoaded data from {survey_file}")

    # Free text entered for the session (e.g. who tested and when), if present.
    session_free_text = survey_data.get('meta', {}).get('session_free_text', '')
    print(f"Session free text: {session_free_text}")

    # Keep only the questions that have a reference answer for this survey, in question order.
    survey_id = survey_data['survey_id']
    ref_questions = sorted((q for (sid, q) in ref_lookup if sid == survey_id), key=lambda q: int(q[1:]))
    if not ref_questions:
        print(f"No reference entries for survey {survey_id}, skipped")
        continue
    questions = [q for q in ref_questions if q in survey_data['answers']]
    missing = [q for q in ref_questions if q not in survey_data['answers']]
    if missing:
        print(f"Not in survey, skipped: {missing}")

    references = [ref_lookup[(survey_id, q)] for q in questions]
    answers = [survey_data['answers'][q]['answer'] for q in questions]
    print(f"Found {len(answers)} transcribed answers matching the reference.")
    for q, ans in zip(questions, answers):
        print(f"{q}: {ans}")

    surveys.append({"file": survey_file, "references": references, "answers": answers})


Loaded data from survey_46ec4547-5bbe-4b02-9c2a-cbcec7223e2c_2026-09-17_16-00-23.json
Session free text: This is Victor Kitoto filling in the survey on 17th September 2026
Found 10 transcribed answers matching the reference.
Q8: Ndiyo, elia ribu majani na kupunguza mawuno.
Q9: Na eza kukubali kupoteza hadia silimi ya kumi ya mawono.
Q10: Ikiva huwari defu inafika karibu asile miatela tini, intaba deleisha maa aina.
Q11: Iwe na mawuna mengi na iwe inasta emili ukame na wadulu
Q12: Atata ni moja kwaekari kama mbegunin zori.
Q13: Wakati wamaoa na kutengeneza punje
Q14: Ninaweka ya chakula na uza iliobaki na kidobu ni amifugo.
Q15: Nwaapa koko nanguruwe maindi marambili hadi tatu kua wiki.
 Nwaapa koko nanguruwe maindi marambili hadi tatu kua wiki.
Q16: Nina nunua kwa agrovet kwa sababu nina aminiu borawake.
Q17: Na uza kwa fanya bia shara wa ineu kwa sababu wa kukaribu na uli paraka.

Loaded data from survey_4f629181-d1dc-46cf-8e71-d819a4b33662_2026-09-21_14-37-06.json
Session free text:

In [16]:
import jiwer 

transform_words = jiwer.Compose ([
    jiwer.ToLowerCase(),
    jiwer.RemovePunctuation(),
    jiwer.RemoveMultipleSpaces(),
    jiwer.Strip(), 
    jiwer.ReduceToListOfListOfWords()
])

transform_chars = jiwer.Compose ([
    jiwer.ToLowerCase(),
    jiwer.RemovePunctuation(),
    jiwer.RemoveMultipleSpaces(),
    jiwer.Strip(), 
    jiwer.ReduceToListOfListOfChars()
])

def error_rates(references, answers):
    out = jiwer.process_words(references, answers, reference_transform=transform_words, hypothesis_transform=transform_words)
    out2 = jiwer.process_characters(references, answers, reference_transform=transform_chars, hypothesis_transform=transform_chars)
    return out.wer, out2.cer

# Per-survey rates.
for s in surveys:
    wer, cer = error_rates(s["references"], s["answers"])
    print(f"{s['file']}  ({len(s['answers'])} answers)")
    print(f"  WER (normalized): {wer:.3f}   CER (normalized): {cer:.3f}")

# Aggregate over every answer from every survey: total errors / total reference words (or characters).
all_references = [r for s in surveys for r in s["references"]]
all_answers = [a for s in surveys for a in s["answers"]]
wer, cer = error_rates(all_references, all_answers)
print(f"\nAggregate over {len(surveys)} surveys, {len(all_answers)} answers")
print(f"  WER (normalized): {wer:.3f}   CER (normalized): {cer:.3f}")

survey_46ec4547-5bbe-4b02-9c2a-cbcec7223e2c_2026-09-17_16-00-23.json  (10 answers)
  WER (normalized): 0.697   CER (normalized): 0.148
survey_4f629181-d1dc-46cf-8e71-d819a4b33662_2026-09-21_14-37-06.json  (10 answers)
  WER (normalized): 0.659   CER (normalized): 0.120
survey_7db59ab6-e13d-4080-bb07-3842d8b51fe4_2026-09-21_08-45-52.json  (10 answers)
  WER (normalized): 0.727   CER (normalized): 0.120

Aggregate over 3 surveys, 30 answers
  WER (normalized): 0.695   CER (normalized): 0.130


## Reference text and extra speech

Surveys from Sep/17–Sep/21: 30 scored answers (Q8–Q17 from each of the 3 surveys). Reference text comes from `office_testing/expected_answers.json`. It has one entry per survey and question, and each entry links to its recording.

- Numbers are spelled out ("asilimia kumi", "tani moja"), matching how they are spoken.
- Where a question was recorded more than once, the entry points to the last take, by filename timestamp.

### The one known repeat

| Survey | Question | What happened |
|---|---|---|
| 2026-09-17 (`46ec4547`, Victor Kitoto) | Q15 | The respondent gave the answer twice in one recording (39.3 s, vs 8–14 s for the other Q15 takes). The app transcribed both. |

Because the speech really was said twice, the reference for this entry now contains the sentence twice. The repeat is scored as expected speech, not as an ASR error.

### Effect on the error rates

| | WER | CER |
|---|---|---|
| 2026-09-17, old single reference | 0.773 | 0.249 |
| 2026-09-17, doubled reference | 0.697 | 0.148 |
| **Aggregate, old single reference** | **0.720** | **0.163** |
| **Aggregate, doubled reference** | **0.695** | **0.130** |

The two Sep/21 surveys are unchanged: WER 0.659 / CER 0.120 for `4f629181`, and WER 0.727 / CER 0.120 for `7db59ab6`.